architecture/src/androidLibMain/kotlin/observable/Observable.kt

```diff
@@ -0,0 +1,135 @@
+/*
+ Copyright 2021 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+@file:JvmName("AndroidObservable")
+
+package com.splendo.kaluga.architecture.observable
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MediatorLiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.Observer
+import androidx.lifecycle.asLiveData
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.awaitCancellation
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+import kotlin.coroutines.CoroutineContext
+import kotlin.properties.ReadOnlyProperty
+import kotlin.reflect.jvm.isAccessible
+
+actual interface WithState<T> {
+    actual val stateFlow:StateFlow<T>
+    actual val valueDelegate: ReadOnlyProperty<Any?, T>
+    val liveData: LiveData<T>
+        get()= stateFlow.asLiveData()
+}
+
+val <R:T, T, OO: ObservableOptional<R>> BasicSubject<R, T, OO>.liveDataObserver: Observer<T>
+    get() = Observer<T> { this.post(it) }
+
+private fun <B, R:T, T, OO: ObservableOptional<R>> B.mutableLiveData(): MutableLiveData<R> where B : BasicSubject<R, T, OO>, B : WithMutableState<R> {
+    val mediatorLiveData = MediatorLiveData<R>()
+    mediatorLiveData.addSource(stateFlow.asLiveData()) { value ->
+        mediatorLiveData.postValue(value)
+    }
+
+    return mediatorLiveData
+}
+
+fun <T> LiveData<T>.observeOnCoroutine(
+    coroutineScope: CoroutineScope,
+    coroutineContext: CoroutineContext = Dispatchers.Main.immediate,
+    observer: Observer<T>
+) {
+    coroutineScope.launch(coroutineContext) {
+        observeForever(observer)
+        awaitCancellation()
+    }.invokeOnCompletion {
+        removeObserver(observer)
+    }
+}
+
+actual abstract class BaseSubject<R:T, T, OO : ObservableOptional<R>> actual constructor(
+    observation: Observation<R, T, OO>,
+    stateFlowToBind:()-> StateFlow<R?>
+): AbstractBaseSubject<R, T, OO>(observation, stateFlowToBind) {
+
+    private lateinit var coroutineScope: CoroutineScope
+    val mutableLiveData: MutableLiveData<R> by lazy {
+        createLiveData().also {
+            if (this::coroutineScope.isInitialized) {
```

Comment:
The purpose of `lateinit` is to assume a var should always have been set except in error, but here it is a matter of which method is called first having different behaviour, so you can just use an optional, especially since it's private.